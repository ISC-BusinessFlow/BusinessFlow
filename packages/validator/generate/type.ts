export interface Error {
  print: () => void;
  getError: () => {
    message: string;
    level: string;
    relation: {
      from?: TaskObject;
      to?: TaskObject;
      path?: PathObject;
    };
  };
}

export type Result = {
  isValid: boolean;
  errors: Error[];
};

export type TaskObject = {
  name: string;
  id: number;
};

export type PathObject = {
  name: string;
  id: number;
};
