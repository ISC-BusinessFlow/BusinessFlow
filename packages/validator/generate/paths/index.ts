// Code generated by validator, DO NOT EDIT

export * from './path1';
export * from './path2';
