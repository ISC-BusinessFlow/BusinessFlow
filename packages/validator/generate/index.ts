/* eslint-disable @typescript-eslint/ban-ts-comment */
import { PathRule, RelationRule, TaskRule } from './base';
// @ts-ignore
import * as paths from './paths';
// @ts-ignore
import * as tasks from './tasks';
import { Error, Result } from './type';

export function appendError(errors: Error[], result: Result) {
  if (!result.isValid) {
    errors.push(...result.errors);
  }
}

export function validateRelation({
  fromId,
  pathId,
  toId,
}: {
  fromId: number;
  pathId: number;
  toId: number;
}): Result {
  // @ts-ignore
  const fromTask: TaskRule = tasks[`task${fromId}`];
  // @ts-ignore
  const path: PathRule = paths[`path${pathId}`];
  // @ts-ignore
  const toTask: TaskRule = tasks[`task${toId}`];

  const errors: Error[] = [];

  appendError(errors, fromTask.validateNextTask(toId));
  appendError(errors, fromTask.validateNextPath(pathId));

  appendError(errors, path.validateNextTask(toId));
  appendError(errors, path.validatePreviousTask(fromId));

  appendError(errors, toTask.validatePreviousTask(fromId));
  appendError(errors, toTask.validatePreviousPath(pathId));

  return {
    isValid: errors.length === 0,
    errors,
  };
}

export function getTaskRule(id: number): RelationRule {
  // @ts-ignore
  const task: TaskRule = tasks[`task${id}`];

  return task.getRule();
}

export function getPathRule(id: number): RelationRule {
  // @ts-ignore
  const path: PathRule = paths[`path${id}`];

  return path.getRule();
}
