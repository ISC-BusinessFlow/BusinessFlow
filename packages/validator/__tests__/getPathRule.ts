import { getPathRule } from '../generate';

describe('getPathRule', () => {
  it('遷移', () => {
    const result = getPathRule(1);
    expect(result).toMatchSnapshot();
  });

  it('通信', () => {
    const result = getPathRule(2);
    expect(result).toMatchSnapshot();
  });
});
