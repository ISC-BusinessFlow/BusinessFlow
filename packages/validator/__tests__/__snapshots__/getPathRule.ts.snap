// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPathRule 通信 1`] = `
Object {
  "from": Object {
    "allowPaths": Array [
      Object {
        "id": 1,
        "name": "遷移",
      },
      Object {
        "id": 2,
        "name": "通信",
      },
    ],
    "allowTasks": Array [
      Object {
        "id": 2,
        "name": "入出力",
      },
      Object {
        "id": 4,
        "name": "システムからの出力",
      },
      Object {
        "id": 5,
        "name": "データストア",
      },
      Object {
        "id": 7,
        "name": "受領",
      },
    ],
    "denyPaths": Array [],
    "denyTasks": Array [
      Object {
        "id": 1,
        "name": "トリガ",
      },
      Object {
        "id": 3,
        "name": "システムからの入力",
      },
      Object {
        "id": 6,
        "name": "処理",
      },
    ],
  },
  "to": Object {
    "allowPaths": Array [
      Object {
        "id": 1,
        "name": "遷移",
      },
      Object {
        "id": 2,
        "name": "通信",
      },
    ],
    "allowTasks": Array [
      Object {
        "id": 1,
        "name": "トリガ",
      },
      Object {
        "id": 2,
        "name": "入出力",
      },
      Object {
        "id": 3,
        "name": "システムからの入力",
      },
      Object {
        "id": 4,
        "name": "システムからの出力",
      },
      Object {
        "id": 5,
        "name": "データストア",
      },
      Object {
        "id": 6,
        "name": "処理",
      },
      Object {
        "id": 7,
        "name": "受領",
      },
    ],
    "denyPaths": Array [],
    "denyTasks": Array [],
  },
}
`;

exports[`getPathRule 遷移 1`] = `
Object {
  "from": Object {
    "allowPaths": Array [
      Object {
        "id": 1,
        "name": "遷移",
      },
      Object {
        "id": 2,
        "name": "通信",
      },
    ],
    "allowTasks": Array [
      Object {
        "id": 1,
        "name": "トリガ",
      },
      Object {
        "id": 2,
        "name": "入出力",
      },
      Object {
        "id": 3,
        "name": "システムからの入力",
      },
      Object {
        "id": 4,
        "name": "システムからの出力",
      },
      Object {
        "id": 5,
        "name": "データストア",
      },
      Object {
        "id": 6,
        "name": "処理",
      },
      Object {
        "id": 7,
        "name": "受領",
      },
    ],
    "denyPaths": Array [],
    "denyTasks": Array [],
  },
  "to": Object {
    "allowPaths": Array [
      Object {
        "id": 1,
        "name": "遷移",
      },
      Object {
        "id": 2,
        "name": "通信",
      },
    ],
    "allowTasks": Array [
      Object {
        "id": 1,
        "name": "トリガ",
      },
      Object {
        "id": 2,
        "name": "入出力",
      },
      Object {
        "id": 3,
        "name": "システムからの入力",
      },
      Object {
        "id": 4,
        "name": "システムからの出力",
      },
      Object {
        "id": 5,
        "name": "データストア",
      },
      Object {
        "id": 6,
        "name": "処理",
      },
      Object {
        "id": 7,
        "name": "受領",
      },
    ],
    "denyPaths": Array [],
    "denyTasks": Array [],
  },
}
`;
